attempt to re-open an already-closed object (1.3.1)

I'm getting an error once in a while in an AsyncTask:

`compile 'com.couchbase.lite:couchbase-lite-android:1.3.1'`

```
Fatal Exception: java.lang.RuntimeException: An error occured while executing doInBackground()
       at android.os.AsyncTask$3.done(AsyncTask.java:300)
       at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:355)
       at java.util.concurrent.FutureTask.setException(FutureTask.java:222)
       at java.util.concurrent.FutureTask.run(FutureTask.java:242)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
       at java.lang.Thread.run(Thread.java:841)
Caused by java.lang.IllegalStateException: attempt to re-open an already-closed object: SQLiteDatabase: /data/data/com.app/files/user.cblite2/db.sqlite3
       at com.couchbase.lite.internal.database.sqlite.SQLiteClosable.acquireReference(SourceFile:60)
       at com.couchbase.lite.internal.database.sqlite.SQLiteDatabase.beginTransaction(SourceFile:497)
       at com.couchbase.lite.internal.database.sqlite.SQLiteDatabase.beginTransaction(SourceFile:410)
       at com.couchbase.lite.storage.SQLiteStorageEngineBase.beginTransaction(SourceFile:150)
       at com.couchbase.lite.store.SQLiteStore.beginTransaction(SourceFile:2309)
       at com.couchbase.lite.store.SQLiteStore.add(SourceFile:1412)
       at com.couchbase.lite.Database.put(SourceFile:1844)
       at com.couchbase.lite.Database.putRevision(SourceFile:1794)
       at com.couchbase.lite.Database.putRevision(SourceFile:1785)
       at com.couchbase.lite.Document.putProperties(SourceFile:523)
       at com.couchbase.lite.Document.putProperties(SourceFile:350)
       at com.app.DevUtilities.Cache.Cache.store(SourceFile:76)
       at com.app.DevUtilities.Cache.CacheObject.apply(SourceFile:46)
       at com.app.Models.Account.Account.save(SourceFile:155)
       at com.app.Models.Account.Account.setOrUpdateAccount(SourceFile:191)
       at com.app.Models.Account.Account.loadApiAccountIntoDb(SourceFile:301)
       at com.app.Fragments.ActivityStream.Controller.ActivityStreamController$ActivityTask.doInBackground(SourceFile:244)
       at com.app.Fragments.ActivityStream.Controller.ActivityStreamController$ActivityTask.doInBackground(SourceFile:215)
       at android.os.AsyncTask$2.call(AsyncTask.java:288)
       at java.util.concurrent.FutureTask.run(FutureTask.java:237)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
       at java.lang.Thread.run(Thread.java:841)
```

Note to Self:

Device: Nexus 9 - Android N
CBL Version: 1.3.1

```
--------- beginning of crash
09-30 15:31:42.719 9997-10251/com.example.tristanrichard.coucbasesample E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #1
        Process: com.example.tristanrichard.coucbasesample, PID: 9997
        java.lang.RuntimeException: An error occurred while executing doInBackground()
            at android.os.AsyncTask$3.done(AsyncTask.java:318)
            at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
            at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
            at java.util.concurrent.FutureTask.run(FutureTask.java:242)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
            at java.lang.Thread.run(Thread.java:761)
         Caused by: java.lang.NullPointerException: Attempt to invoke interface method 'void com.couchbase.lite.storage.SQLiteStorageEngine.setTransactionSuccessful()' on a null object reference
            at com.couchbase.lite.store.SQLiteStore.endTransaction(SQLiteStore.java:2358)
            at com.couchbase.lite.store.SQLiteStore.add(SQLiteStore.java:1607)
            at com.couchbase.lite.Database.put(Database.java:1844)
            at com.couchbase.lite.Database.putRevision(Database.java:1794)
            at com.couchbase.lite.Database.putRevision(Database.java:1785)
            at com.couchbase.lite.Document.putProperties(Document.java:523)
            at com.couchbase.lite.Document.putProperties(Document.java:350)
            at com.example.tristanrichard.coucbasesample.MainActivity.store(MainActivity.java:54)
            at com.example.tristanrichard.coucbasesample.MainActivity$ActivityTask.doInBackground(MainActivity.java:91)
            at com.example.tristanrichard.coucbasesample.MainActivity$ActivityTask.doInBackground(MainActivity.java:73)
            at android.os.AsyncTask$2.call(AsyncTask.java:304)
            at java.util.concurrent.FutureTask.run(FutureTask.java:237)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133) 
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607) 
            at java.lang.Thread.run(Thread.java:761) 
```

--- a/src/androidTest/java/com/couchbase/lite/multithreads/MultiThreadsTest.java
+++ b/src/androidTest/java/com/couchbase/lite/multithreads/MultiThreadsTest.java
@@ -14,6 +14,7 @@
 package com.couchbase.lite.multithreads;
 
 import com.couchbase.lite.CouchbaseLiteException;
+import com.couchbase.lite.CouchbaseLiteRuntimeException;
 import com.couchbase.lite.Database;
 import com.couchbase.lite.Document;
 import com.couchbase.lite.Emitter;
@@ -31,6 +32,8 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 
 /**
@@ -366,5 +369,68 @@ public class MultiThreadsTest extends LiteTestCaseWithDB {
         QueryEnumerator rows = query.run();
         assertEquals(expectCount * 500, gotVu.getLastSequenceIndexed());
         assertEquals(expectCount * 500, rows.getCount());
+    }
+
+    public void testCloseDBDuringInsertions() throws CouchbaseLiteException {
+        if (!multithreadsTestsEnabled())
+            return;
+
+        final Database db = manager.getDatabase("multi-thread-close-db");
+
+        final CountDownLatch latch = new CountDownLatch(50);
+
+        Thread insertThread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                for (int i = 0; i < 100; i++) {
+                    store(i);
+                    latch.countDown();
+                }
+            }
+        });
+
+        Thread closeThread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    assertTrue(latch.await(30, TimeUnit.SECONDS));
+                } catch (InterruptedException e) {
+                    fail(e.toString());
+                }
+                db.close();
+            }
+        });
+
+        closeThread.start();
+        insertThread.start();
+
+        try {
+            closeThread.join();
+        } catch (InterruptedException e) {
+            Log.e(TAG, "Error in closeThread. ", e);
+        }
+        try {
+            insertThread.join();
+        } catch (InterruptedException e) {
+            Log.e(TAG, "Error in insertThread. ", e);
+        }
+
+        db.delete();
+    }
+
+    private void store(int i) {
+        try {
+            Database db = manager.getDatabase("multi-thread-close-db");
+            String id = String.format(Locale.ENGLISH, "ID_%05d", i);
+            Document doc = db.getDocument(id);
+            Map<String, Object> props = new HashMap<String, Object>();
+            props.put("value", i);
+            doc.putProperties(props);
+        } catch (CouchbaseLiteRuntimeException re) {
+            Log.e(TAG, "Error in store(): %s", re.getMessage());
+        } catch (CouchbaseLiteException e) {
+            Log.e(TAG, "Error in store()", e);
+            fail("Error in store()");
+        }
     }
 }

--- a/src/androidTest/java/com/couchbase/lite/store/SQLiteStoreTest.java
+++ b/src/androidTest/java/com/couchbase/lite/store/SQLiteStoreTest.java
@@ -18,6 +18,7 @@ import com.couchbase.lite.LiteTestCaseWithDB;
 import com.couchbase.lite.SavedRevision;
 import com.couchbase.lite.UnsavedRevision;
 
+import java.lang.reflect.Method;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.atomic.AtomicBoolean;
@@ -38,7 +39,9 @@ public class SQLiteStoreTest extends LiteTestCaseWithDB {
         if(!isSQLiteDB())
             return;
 
-        SQLiteStore store = (SQLiteStore) database.getStore();
+        Method m = database.getClass().getDeclaredMethod("getStore");
+        m.setAccessible(true);
+        SQLiteStore store = (SQLiteStore)m.invoke(database);
 
         Map<String, Object> properties = new HashMap<String, Object>();
         properties.put("testName", "testCreateRevisions");

